ls1intum/Artemis
# `Programming exercises`: Make Jenkins pipelines more flexible by storing the pipeline definition in Artemis

### Checklist
#### General


#### Server

### Motivation and Context


This issue explains the underlying motivation in detail: #5726 (Closes #5726)
Note that this pull request only affects the Jenkins/GitLab configuration of Artemis. 

In some cases, custom build plans are required in order to align the Artemis Jenkins build plans with the educational institution's previously used proprietary testing mechanism. While this has already been possible before, the process was fairly complicated and cumbersome. In order to make editing build plans easier, this PR sets the foundation for a per-exercise build plan editor in the backend. 

### Description

Previously, there were multiple Jenkinsfiles for different exercise configurations. Those Jenkinsfiles all contained the same information regarding the Artemis instance, but differed in the specific build steps based on programming language and specific settings, such as static code analysis. With this PR, the common part of those Jenkinsfiles is extracted and put into only one Jenkinsfile that now serves as base template for all possible exercise configurations. The instance specific replacements for this common template are made by the ```JenkinsBuildPlanCreator``` class.  
Programming exercise specific build steps are stored in ```pipeline.groovy``` files, which are now stored in the programming language directories. The replacements for those pipelines are made by the ```JenkinsPipelineScriptCreator``` class. This build plan pipeline is now stored in the database as an attribute of an exercise, which allows for custom build plans for every exercise.


### Steps for Testing

Prerequisites:
- 1 Editor, Instructor or Admin
- 1 Course
- **TS2**, server with a **GitLab/Jenkins** setup

#### How to test the build plan creation:

Testing this via #6414 is easier since that PR makes the build plans available to edit in the Artemis UI.

Steps in Artemis:
1. Log in to Artemis
2. Navigate to Course Management
3. Click ```Exercises``` on the bottom left of the course
4. Create a new Programming Exercise, it can be configured in any way.
   - The different combinations of programming languages and feature-sets (testwise coverage, static code analysis, stepwise testing) should be tested in multiple exercises.
5. Select the generated programming exercise and scroll down to the **Solution Result** section
6. If static code analysis was enabled, there should be three warnings when you click on the result.

Steps in Jenkins:
1. Log in to Jenkins (https://jenkins-test.artemis.in.tum.de/), log in with the same user as in artemis-test2
2. Select the newly created exercise
3. Confirm that the base build fails and the solutions build works (can be checked in Artemis, too, on the exercise details page)
4. Select either the base or solution build
5. Select ```Configure``` on the menu on the left
6. Scroll down to the ```Pipeline``` section
7. Confirm that the replacements have been made corresponding to the settings that were chosen during the exercise creation (i.e. there should be no ```#``` symbol left in the file).

#### How to test the build plan re-creation of an existing exercise:
Steps in Artemis:
1. Log in to Artemis
2. Navigate to Course Management
3. Click ```Exercises``` on the bottom left of the course
4. Select the previously tested exercise
5. Click ```Build Plan Reset```
6. Confirm that there is no error shown in the UI (currently there is no way to edit the build plan, so the old build plan gets replaced with an identical new one)

Steps in Jenkins:
1. Log in to Jenkins
2. Select the exercise from before
3. Confirm that both the ```BASE``` and ```SOLUTION``` plan are re-built
4. Confirm that the results are the same as they were before the build plan has been reset

### Review Progress






#### Code Review
#### Manual Tests (also covered by manual tests on #6414)

### Test Coverage




| Class/File | Branch % | Line % |
|------------|-------:|-----:|
| `JenkinsBuildPlanService` | 82 | 93 |
| `JenkinsPipelineScriptCreator` | 92 | 94 |
| `JenkinsBuildPlanCreator` | 100 | 96 |
